kernel/kernel-api/src/main/java/io/delta/kernel/internal/skipping/DataSkippingUtils.java

```diff
@@ -345,17 +353,49 @@ private static DataSkippingPredicate constructBinaryDataSkippingPredicate(
             String exprName,
             Column col,
             Literal lit) {
+
+        return constructBinaryDataSkippingPredicate(
+            exprName, col, lit, Collections.singleton(col));
+    }
+
+    /**
+     * Constructs a {@link DataSkippingPredicate} for a binary predicate expression with a left
+     * column expression and a right expression of type {@link Literal}.
+     */
+    private static DataSkippingPredicate constructBinaryDataSkippingPredicate(
+        String exprName,
+        Expression colExpr,
+        Literal lit,
+        Set<Column> referencedCols) {
         return new DataSkippingPredicate(
             exprName,
-            Arrays.asList(col, lit),
-            new HashSet<Column>(){
-                {
-                    add(col);
-                }
-            }
+            Arrays.asList(colExpr, lit),
+            referencedCols
         );
     }
 
+    /**
+     * Adjusts the timestamp column for max comparisons by adding 1 microsecond if needed.
+     * This adjustment compensates for truncation to milliseconds in JSON serialization.
+     *
+     * @param col The column to potentially adjust.
+     * @param schemaHelper The schema helper to use for column information.
+     * @param addMillisecond Whether to add 1 millisecond to the column.
+     * @return The adjusted or original column expression.
+     */
+    private static Tuple2<Expression, Set<Column>> adjustTimestampForMaxColumnIfNeeded(
+        Column col, StatsSchemaHelper schemaHelper, boolean addMillisecond) {
+        DataType colType = schemaHelper.getDataTypeFromLogicalColumn(col);
+        Column maxColumn = schemaHelper.getMaxColumn(col);
+        if (colType instanceof TimestampType && addMillisecond) {
```

Comment:
can this work for `TimestampNTZType` column too?

Reply:
Yup, it can. Updating to handle it